exec/java-exec/src/main/java/org/apache/drill/exec/store/avro/AvroBatchReader.java

```diff
@@ -0,0 +1,368 @@
+import org.apache.avro.generic.GenericFixed;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.mapred.FsInput;
+import org.apache.avro.util.Utf8;
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.exec.physical.impl.scan.file.FileScanFramework;
+import org.apache.drill.exec.physical.impl.scan.framework.ManagedReader;
+import org.apache.drill.exec.physical.resultSet.ResultSetLoader;
+import org.apache.drill.exec.physical.resultSet.RowSetLoader;
+import org.apache.drill.exec.record.metadata.ColumnMetadata;
+import org.apache.drill.exec.record.metadata.TupleMetadata;
+import org.apache.drill.exec.util.ImpersonationUtil;
+import org.apache.drill.exec.vector.accessor.ArrayWriter;
+import org.apache.drill.exec.vector.accessor.DictWriter;
+import org.apache.drill.exec.vector.accessor.ObjectWriter;
+import org.apache.drill.exec.vector.accessor.ScalarWriter;
+import org.apache.drill.exec.vector.accessor.TupleWriter;
+import org.apache.drill.shaded.guava.com.google.common.base.Charsets;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.FileSplit;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.joda.time.DateTimeConstants;
+import org.joda.time.Period;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+import java.security.PrivilegedExceptionAction;
+import java.util.List;
+import java.util.Map;
+
+public class AvroBatchReader implements ManagedReader<FileScanFramework.FileSchemaNegotiator> {
+
+  private static final Logger logger = LoggerFactory.getLogger(AvroBatchReader.class);
+
+  // currently config is unused but maybe used later
+  private final AvroReaderConfig config;
+
+  private Path filePath;
+  private long endPosition;
+  private DataFileReader<GenericContainer> reader;
+  private ResultSetLoader loader;
+  // re-use container instance
+  private GenericContainer container = null;
+
+  public AvroBatchReader(AvroReaderConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public boolean open(FileScanFramework.FileSchemaNegotiator negotiator) {
+    FileSplit split = negotiator.split();
+    filePath = split.getPath();
+
+    // Avro files are splittable, define reading start / end positions
+    long startPosition = split.getStart();
+    endPosition = startPosition + split.getLength();
+
+    logger.debug("Processing Avro file: {}, start position: {}, end position: {}",
+      filePath, startPosition, endPosition);
+
+    reader = prepareReader(split, negotiator.fileSystem(),
+      negotiator.userName(), negotiator.context().getFragmentContext().getQueryUserName());
+
+    logger.debug("Avro file schema: {}", reader.getSchema());
+    TupleMetadata schema = AvroSchemaUtil.convert(reader.getSchema());
```

Comment:
Here we are converting the Avro schema to a Drill schema, which makes sense. If we scan a directory with multiple files, they should all have the same schema (or an evolved schema: `(A, B) --> (A,B,C) --> (B,C)`.

In the future, it would be handy to infer the schema a plan time, then pass the resulting schema to each reader so we don't have to repeat identical work in each of perhaps hundreds of readers.

Reply:
Sounds good, is there examples for EVF where schema is inferred at plan time? When doing this implementation, I looked at existing EVF implementations and mostly they inferred schema in reader.

Reply:
I actually don't have a good example in mind. Planner-side schema planning is not really an EVF-specific thing. We used an external schema for CSV files using your provisioned schema mechanism. There is the new metadata mechanism that was recently added. We do partition and row group pruning for Parquet using external metadata.

So, planner-side schema analysis is probably something new we'd want to add. I believe that Avro, when used for RPC, defines the schema separate from the data. Is something like that available when Avro is used as a file format?

For now, perhaps we can just do the schema planning at open time for each file. Then, if that code just does Avro --> ColumnMetadata translation, the code can be moved earlier in the process if we find a way to obtain the schema earlier.

Reply:
When reading Avro files we obtain schema from each file. 
I believe we might have list of files to read at planning time (that what we have for Parquet where we read the footers). We can open first file read the schema, convert it and then provide it when reading all files but of course we would have to check each file is compatible with such schema. Though from the EVF code, I am not sure where such approach can be added.